src/InstrumentationEngine.Attach/Program.cs

```diff
@@ -0,0 +1,170 @@
+            {
+                process = Process.GetProcessById(processId);
+            }
+            catch (Exception ex)
+            {
+                WriteErrorJson(ex.Message);
+                return;
+            }
+
+            // This executable should be in the [Root]\Tools\Attach directory. Get the root directory
+            // and build back to the instrumentation engine file path.
+
+            // CONSIDER: Is there a better way to get the engine path instead of building the path
+            // relative to the current executable? For example, requiring the root path of the engine
+            // to be passed as a parameter (which increases the difficulty of using this executable).
+
+            string? attachDirectory = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
+            if (null == attachDirectory || !Directory.Exists(attachDirectory))
+            {
+                WriteErrorJson(Invariant($"Directory '{attachDirectory}' does not exist."));
+                return;
+            }
+
+            string? toolsDirectory = Path.GetDirectoryName(attachDirectory);
+            if (null == toolsDirectory || !Directory.Exists(toolsDirectory))
+            {
+                WriteErrorJson(Invariant($"Directory '{toolsDirectory}' does not exist."));
+                return;
+            }
+
+            string? rootDirectory = Path.GetDirectoryName(toolsDirectory);
+            if (null == rootDirectory || !Directory.Exists(rootDirectory))
+            {
+                WriteErrorJson(Invariant($"Directory '{rootDirectory}' does not exist."));
+                return;
+            }
+
+            string enginePath;
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+            {
+                enginePath = Path.Combine(rootDirectory, "Instrumentation64", "libInstrumentationEngine.so");
+            }
+            else
+            {
+                bool isWow64;
+                try
+                {
+                    // isWow64 is true for 32-bit applications running on a 64-bit OS
+                    if (!NativeMethods.IsWow64Process(process.Handle, out isWow64))
+                    {
+                        Exception? ex = Marshal.GetExceptionForHR(Marshal.GetHRForLastWin32Error());
+                        if (null != ex)
+                        {
+                            WriteErrorJson(Invariant($"Failed to check process bitness: {ex.Message}"));
+                        }
+                        else
+                        {
+                            WriteErrorJson(Invariant($"Failed to check process bitness."));
+                        }
+                    }
+                }
+                catch (Win32Exception ex)
+                {
+                    WriteErrorJson(Invariant($"Failed to check process bitness (code: {ex.NativeErrorCode}): {ex.Message}"));
+                    return;
+                }
+
+                enginePath = isWow64 ?
```

Comment:
>isWow64 [](start = 29, length = 7)

IsWow64 returns false on 32-bit Windows. Do we care about that?

Reply:
The Proxy uses this to determine whether to use ProgramFiles or ProgramFiles(x86) which is nice since ProgramFiles corresponds to the bitness of the machine.

Could we also use agent's bitness to determine whether we're running on 32-bit or 64-bit process if IsWow64 is false?

---
In reply to: [384696104](https://github.com/microsoft/CLRInstrumentationEngine/pull/212#discussion_r384696104) [](ancestors = 384696104)

Reply:
The correct way to do this is:

bool is64Bit = (sizeof(IntPtr) == 8) && !isWow64
enginePath = (is64Bit) ? ....


---
In reply to: [384696104](https://github.com/microsoft/CLRInstrumentationEngine/pull/212#discussion_r384696104) [](ancestors = 384696104)

Reply:
Oh, but you also have to take into account whether *this process* is 64 bit:
bool thisIs64 = (sizeof(IntPtr) == 8) && !thisIsWow;
bool is64BitProc = (thisIs64 && !isWow) || ((!thisIs64 && thisIsWow) && !isWow);

Or something of the like :-)


---
In reply to: [384715740](https://github.com/microsoft/CLRInstrumentationEngine/pull/212#discussion_r384715740) [](ancestors = 384715740,384696104)

Reply:
This code is specifically checking if the target process (not the current process) is 32 bit or not. Anything the current process indicates should not be considered.

---
In reply to: [384718155](https://github.com/microsoft/CLRInstrumentationEngine/pull/212#discussion_r384718155) [](ancestors = 384718155,384715740,384696104)

Reply:
Currently, CLRIE is only running on 64 bit OSes in production diagnostics scenarios. But I can add some code to check the OS bitness in the case that isWow64 is false.

---
In reply to: [384713327](https://github.com/microsoft/CLRInstrumentationEngine/pull/212#discussion_r384713327) [](ancestors = 384713327,384696104)

Reply:
Doesn't matter. You can't tell if the target process is 32bit without knowing if *this* process is a WOW process or not. IsWOW64 tells you "the given process is a 32bit process on a 64bit machine". If you are running on a 32 bit machine IsWOW64 will always return false. So, if IsWOW64 returns false in this case, that means one of two things:

The target process is 64bit OR the target process is 32bit running on a 32bit OS.

To determine which of those is the case, you have to check what the OS architecture is. There is no way to check that. You have to check what the bitness of your current process is. The way to do that is to check the size of a pointer.

But wait, there's more!

The size of the pointer is not good enough. If it is 4bytes, then there are two options:

1) The OS architecture is 32bit
2) The OS architecture is 64bit, and this process is running under WOW64.

To figure out that, you need to check whether *this* process is WOW64, hence the complicated Boolean logic above.

---
In reply to: [384728717](https://github.com/microsoft/CLRInstrumentationEngine/pull/212#discussion_r384728717) [](ancestors = 384728717,384718155,384715740,384696104)

Reply:
I had a larger comment about this, but then I saw that you are using OsArchitecture now. I didn't know that was available in .NET.

---
In reply to: [384728717](https://github.com/microsoft/CLRInstrumentationEngine/pull/212#discussion_r384728717) [](ancestors = 384728717,384718155,384715740,384696104)